src/hotspot/share/opto/type.cpp

```diff
@@ -2406,6 +2412,14 @@ const TypeVect* TypeVect::make(const Type *elem, uint length) {
   return NULL;
 }
 
+const TypeVect *TypeVect::makemask(const BasicType elem_bt, uint length) {
+  if (Matcher::has_predicated_vectors()) {
+    return (TypeVect*)(new TypeVectMask(get_const_basic_type(T_BOOLEAN), MaxVectorSize))->hashcons();
```

Comment:
Do you mean the `TypeVectMask` only saves the `T_BOOLEAN` as the basic type for all the mask type?

Reply:
Yes,  that is correct for targets which do not support predicate registers flow differs and a TypeVect matches with type of vector register.

Reply:
Thanks! Actually that's correct for AVX-512. However, unlike AVX-512 K registers, SVE predicate needs the real basic type as a reference when controlling the vector operations. I'm worried that this is not friendly for SVE.

Reply:
The main reason is that the lane size of SVE predicate depends on the vector lane type, i.e `1-mask-bit per one byte` while `2-mask-bit per one short` and so on. So most of the predicate operations depend on the vector element basic type. We uses the `CNTP` as an example which computes the active numbers in a predicate register:  CNTP` <Xd>, <Pg>, <Pn>.B` is different from `CNTP <Xd>, <Pg>, <Pn>.H`. Another usage of the real basic type is the API `VectorMask.toVector()`. When creating a vector from a mask, we need the real element type (T_BYTE is enough).

Reply:
Ok, so it's a mask bit per vector byte then. It looks similar to how AVX/AVX2 represents masks (w/o predicate registers, but reusing wide vector registers).

Reply:
Yeah, it's a mask bit per byte.